### PR TITLE
Feature custom losses

### PR DESCRIPTION
**Pull request by jay3p**
- Implementing custom loss functions
  - KL-Divergence with normal distribution as the true distribution
  - Expected distance (weighted sum of distance of each value and their probability)
  - Sum of the value probabilities that is lower than the true a specified value
- Adding custom losses to use for each feature
- Option to enable all custom losses (enable_custom_losses)

### CODE REVIEW

File: src/groove_panda/models/tf_custom/feature_losses.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+@register_keras_serializable(package="Custom", name="Velocity")
+class Velocity(Loss):
+    """
+    Uses soft KL divergence (normal distributed)
+    """
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+    def call(self, y_true, y_pred):
+        soft_kl = NormalDistributedCategorialKLDivergence(sigma=2, epsilon=1e-6)
+        return soft_kl(y_true, y_pred)
```
Comment (Eaglewolf13) [inline src/groove_panda/models/tf_custom/feature_losses.py:34]: I believe distance would be a good idea here too!

File: src/groove_panda/models/tf_custom/feature_losses.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+@register_keras_serializable(package="Custom", name="Duration")
+class Duration(Loss):
+    """
+    Uses soft KL divergence (normal distributed)
+    """
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+    def call(self, y_true, y_pred):
+        soft_kl = NormalDistributedCategorialKLDivergence(sigma=0.5, epsilon=1e-6)
+        return soft_kl(y_true, y_pred)
```
Comment (Eaglewolf13) [inline src/groove_panda/models/tf_custom/feature_losses.py:45]: same here with distance

File: src/groove_panda/models/tf_custom/math_utils.py [hunk 3]
```diff
@@ -7,3 +8,35 @@ def nuclear_norm(matrix: tf.Tensor) -> tf.Tensor:
...
+def expected_mse(distribution: tf.Tensor, target: tf.Tensor) -> tf.Tensor:
+    index_range = tf.cast(tf.shape(distribution)[0], tf.float32)
+    max_possible_distance = index_range - 1
+
+    indexes = tf.cast(tf.range(start = 0, limit = index_range), tf.float32)
+    target = tf.cast(target, tf.float32)
+    difference = indexes - target
+    normalized_difference = difference / max_possible_distance
+    mse_distances = normalized_difference ** 2
+    expected_mse_value = tf.reduce_sum(mse_distances * distribution)
```
Comment (Eaglewolf13) [inline src/groove_panda/models/tf_custom/math_utils.py:21]: Are we not multiplying the distribution with 0 for the probability of the correct output?

Comment (Eaglewolf13) [reply src/groove_panda/models/tf_custom/math_utils.py:21]: nevermind, I get it, we want to punish things that were not correct based on their distance and probability, and for this, the correct probability doesn't matter

File: src/groove_panda/models/tf_custom/math_utils.py [hunk 4]
```diff
@@ -7,3 +8,35 @@ def nuclear_norm(matrix: tf.Tensor) -> tf.Tensor:
...
+
+    return expected_mse_value  # noqa: RET504
+
+def sum_under_range(distribution: tf.Tensor, limit: tf.Tensor) -> tf.Tensor:
```
Comment (Eaglewolf13) [inline src/groove_panda/models/tf_custom/math_utils.py:25]: Shouldn't we flip these? So like limit first and distribution second? Because distribution is the model prediction and not y_true?

File: src/groove_panda/models/tf_custom/losses.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+@register_keras_serializable(package="Custom", name="CategoricalExpectedMSE")
+class SumUnderRange(Loss):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def call(self, y_true: tf.Tensor, y_pred: tf.Tensor):
+        return tf.map_fn(
+            lambda x: math_utils.sum_under_range(x[0], x[1]),
+            (y_pred, y_true),
```
Comment (Eaglewolf13) [inline src/groove_panda/models/tf_custom/losses.py:102]: aha disregard my "flipping" comment, I see that you do it here XD

File: src/groove_panda/models/tf_custom/math_utils.py (hunk 4 above)
Comment (Eaglewolf13) [reply src/groove_panda/models/tf_custom/math_utils.py:25]: fixed

File: src/groove_panda/models/tf_custom/math_utils.py [hunk 6]
```diff
@@ -7,3 +8,35 @@ def nuclear_norm(matrix: tf.Tensor) -> tf.Tensor:
...
+    limit = tf.maximum(tf.cast(limit, tf.int32), 0)
+    under_limit = distribution[:limit]
```
Comment (Eaglewolf13) [inline src/groove_panda/models/tf_custom/math_utils.py:27]: You're punishing going back in time extra on top of the general distance punishment, I see

File: src/groove_panda/models/tf_custom/math_utils.py [hunk 7]
```diff
@@ -7,3 +8,35 @@ def nuclear_norm(matrix: tf.Tensor) -> tf.Tensor:
...
+    return tf.reduce_sum(under_limit)
+
+def generate_normal_distribution_array(sigma, epsilon):
```
Comment (Eaglewolf13) [inline src/groove_panda/models/tf_custom/math_utils.py:30]: Note for self:

sigma is the standard deviation, so how large we want to make a deviation (this implies sigma = 0 means no standard deviation)

epsilon is the tolerance, so we ignore values smaller than epsilon and turn them to 0
